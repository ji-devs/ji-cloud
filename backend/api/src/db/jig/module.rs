use anyhow::Context;
use shared::domain::jig::{
    module::{Module, ModuleBody, ModuleId, ModuleKind, StableModuleId, StableOrUniqueId},
    JigId,
};
use sqlx::PgPool;
use std::cmp;

fn map_module_contents(body: &ModuleBody) -> anyhow::Result<(ModuleKind, serde_json::Value)> {
    let kind = body.kind();

    let body = match body {
        ModuleBody::CardQuiz(body) => serde_json::to_value(body)?,
        ModuleBody::Cover(body) => serde_json::to_value(body)?,
        ModuleBody::ResourceCover(body) => serde_json::to_value(body)?,
        ModuleBody::DragDrop(body) => serde_json::to_value(body)?,
        ModuleBody::Flashcards(body) => serde_json::to_value(body)?,
        ModuleBody::Matching(body) => serde_json::to_value(body)?,
        ModuleBody::MemoryGame(body) => serde_json::to_value(body)?,
        ModuleBody::Poster(body) => serde_json::to_value(body)?,
        ModuleBody::TappingBoard(body) => serde_json::to_value(body)?,
        ModuleBody::Video(body) => serde_json::to_value(body)?,
        ModuleBody::FindAnswer(body) => serde_json::to_value(body)?,
        ModuleBody::Legacy(body) => serde_json::to_value(body)?,

        _ => anyhow::bail!("Unimplemented body kind: {}", kind.as_str()),
    };

    Ok((kind, body))
}

fn transform_response_kind(
    contents: serde_json::Value,
    kind: ModuleKind,
) -> anyhow::Result<ModuleBody> {
    match kind {
        ModuleKind::CardQuiz => Ok(ModuleBody::CardQuiz(serde_json::from_value(contents)?)),
        ModuleKind::Cover => Ok(ModuleBody::Cover(serde_json::from_value(contents)?)),
        ModuleKind::ResourceCover => {
            Ok(ModuleBody::ResourceCover(serde_json::from_value(contents)?))
        }
        ModuleKind::DragDrop => Ok(ModuleBody::DragDrop(serde_json::from_value(contents)?)),
        ModuleKind::Flashcards => Ok(ModuleBody::Flashcards(serde_json::from_value(contents)?)),
        ModuleKind::Matching => Ok(ModuleBody::Matching(serde_json::from_value(contents)?)),
        ModuleKind::Memory => Ok(ModuleBody::MemoryGame(serde_json::from_value(contents)?)),
        ModuleKind::Poster => Ok(ModuleBody::Poster(serde_json::from_value(contents)?)),
        ModuleKind::TappingBoard => Ok(ModuleBody::TappingBoard(serde_json::from_value(contents)?)),
        ModuleKind::Video => Ok(ModuleBody::Video(serde_json::from_value(contents)?)),
        ModuleKind::FindAnswer => Ok(ModuleBody::FindAnswer(serde_json::from_value(contents)?)),
        ModuleKind::Legacy => Ok(ModuleBody::Legacy(serde_json::from_value(contents)?)),

        _ => anyhow::bail!("Unimplemented response kind"),
    }
}

pub async fn create(
    pool: &PgPool,
    parent: JigId,
    body: ModuleBody,
    is_complete: bool,
) -> anyhow::Result<(ModuleId, u16)> {
    let (kind, body) = map_module_contents(&body)?;

    let mut txn = pool.begin().await?;

    let draft_id = sqlx::query!(
        //language=SQL
        r#"
select draft_id from jig where jig.id = $1
"#,
        parent.0,
    )
    .fetch_optional(&mut txn)
    .await?
    .map(|it| it.draft_id);

    let res = sqlx::query!(
        //language=SQL
        r#"
insert into jig_data_module (jig_data_id, kind, contents, index, is_complete)
values ($1, $2, $3, (select count(*) from jig_data_module where jig_data_id = $1), $4)
returning id, "index"
"#,
        draft_id,
        kind as i16,
        body,
        is_complete,
    )
    .fetch_one(&mut txn)
    .await
    .map(|it| (ModuleId(it.id), it.index as u16))
    .map_err(Into::into);

    txn.commit().await?;

    res
}

pub async fn get_live(
    pool: &PgPool,
    parent: JigId,
    lookup: StableOrUniqueId,
) -> anyhow::Result<Option<Module>> {
    let (unique, stable) = (lookup.unique(), lookup.stable());

    let module = sqlx::query!(
        //language=SQL
        r#"
select id          as "id: ModuleId",
       stable_id   as "stable_id: StableModuleId",
       contents    as "body",
       created_at  as "created_at",
       updated_at  as "updated_at",
       kind        as "kind: ModuleKind",
       is_complete as "is_complete"
from jig_data_module
where jig_data_module.id is not distinct from $2
   or (jig_data_id = (select live_id from jig where jig.id = $1) and stable_id is not distinct from $3)
"#,
        parent.0,
        unique.map(|it| it.0),
        stable.map(|it| it.0)
    )
    .fetch_optional(pool)
    .await?;

    let map_response = |body, kind| transform_response_kind(body, kind);

    match module {
        Some(it) => Ok(Some(Module {
            id: it.id,
            stable_id: it.stable_id,
            created_at: it.created_at,
            updated_at: it.updated_at,
            body: map_response(it.body, it.kind).context(anyhow::anyhow!(
                "failed to transform module of kind {:?}",
                it.kind
            ))?,
            is_complete: it.is_complete,
            is_updated: it.created_at < it.updated_at,
        })),
        None => Ok(None),
    }
}

/// FIXME dedup this from live JIG
pub async fn get_draft(
    pool: &PgPool,
    parent: JigId,
    lookup: StableOrUniqueId,
) -> anyhow::Result<Option<Module>> {
    let (unique, stable) = (lookup.unique(), lookup.stable());

    let module = sqlx::query!(
        //language=SQL
        r#"
select id          as "id: ModuleId",
       stable_id   as "stable_id: StableModuleId",
       contents    as "body",
       created_at  as "created_at",
       updated_at  as "updated_at",
       kind        as "kind: ModuleKind",
       is_complete as "is_complete"
from jig_data_module
where jig_data_module.id is not distinct from $2
   or (jig_data_id = (select draft_id from jig where jig.id = $1) and stable_id is not distinct from $3)
"#,
        parent.0,
        unique.map(|it| it.0),
        stable.map(|it| it.0)
    )
        .fetch_optional(pool)
        .await?;

    let map_response = |body, kind| transform_response_kind(body, kind);

    match module {
        Some(it) => Ok(Some(Module {
            id: it.id,
            stable_id: it.stable_id,
            created_at: it.created_at,
            updated_at: it.updated_at,
            body: map_response(it.body, it.kind).context(anyhow::anyhow!(
                "failed to transform module of kind {:?}",
                it.kind
            ))?,
            is_complete: it.is_complete,
            is_updated: it.created_at < it.updated_at,
        })),
        None => Ok(None),
    }
}

pub async fn update(
    pool: &PgPool,
    parent_id: JigId,
    lookup: StableOrUniqueId,
    body: Option<&ModuleBody>,
    new_index: Option<u16>,
    is_complete: Option<bool>,
) -> anyhow::Result<bool> {
    let (unique, stable) = (lookup.unique(), lookup.stable());

    let (kind, body) = match body.map(map_module_contents).transpose()? {
        Some((kind, body)) => (Some(kind), Some(body)),
        None => (None, None),
    };

    let mut txn = pool.begin().await?;

    let draft_id = sqlx::query!(
        //language=SQL
        r#"
select draft_id from jig where jig.id = $1
"#,
        parent_id.0,
    )
    .fetch_optional(&mut txn)
    .await?
    .map(|it| it.draft_id);

    let index = sqlx::query!(
        //language=SQL
        r#"
select index from jig_data_module
where jig_data_module.id is not distinct from $2
   or (jig_data_id = $1 and stable_id is not distinct from $3)
"#,
        draft_id,
        unique.map(|it| it.0),
        stable.map(|it| it.0),
    )
    .fetch_optional(&mut txn)
    .await?;

    let index = match index {
        Some(it) => it.index,
        None => return Ok(false),
    };

    sqlx::query!(
        //language=SQL
        r#"
update jig_data_module
set contents    = coalesce($3, contents),
    kind        = coalesce($4, kind),
    is_complete = coalesce($5, is_complete)
where jig_data_id = $1
  and index = $2
"#,
        draft_id,
        index,
        body.as_ref(),
        kind.map(|it| it as i16),
        is_complete,
    )
    .execute(&mut txn)
    .await?;

    if let Some(new_index) = new_index {
        let new_index = new_index as i16;

        let max_index = sqlx::query!(
            //language=SQL
            r#"select count(*) - 1 as "max_index!" from jig_data_module where jig_data_id = $1"#,
            draft_id
        )
        .fetch_one(&mut txn)
        .await?
        .max_index;

        let new_index = cmp::min(new_index, max_index as i16);

        if new_index < index {
            sqlx::query!(
                //language=SQL
                r#"
update jig_data_module
set
    index = case when index = $2 then $3 else index + 1 end,
    updated_at = now()
where jig_data_id = $1 and index between $3 and $2
"#,
                draft_id,
                index,
                new_index
            )
            .execute(&mut txn)
            .await?;
        } else if new_index > index {
            sqlx::query!(
                //language=SQL
                r#"
update jig_data_module
set
    index = case when index = $2 then $3 else index - 1 end,
    updated_at = now()
where jig_data_id = $1 and index between $2 and $3
"#,
                parent_id.0,
                index,
                new_index
            )
            .execute(&mut txn)
            .await?;
        }
    }

    txn.commit().await?;

    Ok(true)
}

pub async fn delete(pool: &PgPool, parent: JigId, lookup: StableOrUniqueId) -> anyhow::Result<()> {
    let (unique, stable) = (lookup.unique(), lookup.stable());

    let mut txn = pool.begin().await?;

    let draft_id = sqlx::query!(
        //language=SQL
        r#"
select draft_id from jig where jig.id = $1
"#,
        parent.0,
    )
    .fetch_optional(&mut txn)
    .await?
    .map(|it| it.draft_id);

    let idx = sqlx::query!(
        //language=SQL
        r#"
delete
from jig_data_module
where jig_data_module.id is not distinct from $2
   or (jig_data_id = $1 and stable_id is not distinct from $3)
returning index
"#,
        draft_id,
        unique.map(|it| it.0),
        stable.map(|it| it.0),
    )
    .fetch_optional(&mut txn)
    .await?
    .map(|it| it.index);

    if let Some(idx) = idx {
        sqlx::query!(
            //language=SQL
            r#"
update jig_data_module
set index = index - 1
where jig_data_id = $1
  and index > $2
"#,
            draft_id,
            idx,
        )
        .execute(&mut txn)
        .await?;
    }

    txn.commit().await?;

    Ok(())
}
