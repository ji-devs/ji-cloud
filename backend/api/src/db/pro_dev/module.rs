use anyhow::Context;
use shared::domain::{
    module::{Module, ModuleBody, ModuleId, ModuleKind},
    pro_dev::ProDevId,
};
use sqlx::PgPool;
use std::cmp;

pub async fn create(
    pool: &PgPool,
    parent: ProDevId,
    body: ModuleBody,
    is_complete: bool,
) -> anyhow::Result<(ModuleId, u16)> {
    let (kind, body) = ModuleBody::map_module_contents(&body)?;

    let mut txn = pool.begin().await?;

    let draft_id = sqlx::query!(
        //language=SQL
        r#"
select draft_id from pro_dev where pro_dev.id = $1
"#,
        parent.0,
    )
    .fetch_optional(&mut txn)
    .await?
    .map(|it| it.draft_id);

    let res = sqlx::query!(
        //language=SQL
        r#"
insert into pro_dev_data_module (pro_dev_data_id, kind, contents, index, is_complete)
values ($1, $2, $3, (select count(*) from pro_dev_data_module where pro_dev_data_id = $1), $4)
returning id, "index"
"#,
        draft_id,
        kind as i16,
        body,
        is_complete
    )
    .fetch_one(&mut txn)
    .await
    .map(|it| (ModuleId(it.id), it.index as u16))
    .map_err(Into::into);

    txn.commit().await?;

    res
}

pub async fn get_live(pool: &PgPool, id: ModuleId) -> anyhow::Result<Option<Module>> {
    let module = sqlx::query!(
        //language=SQL
        r#"
select cdm.id          as "id!: ModuleId",
       contents    as "body!",
       created_at  as "created_at!",
       updated_at  as "updated_at!",
       kind        as "kind!: ModuleKind",
       is_complete as "is_complete!"
from pro_dev_data_module "cdm"
inner join pro_dev on pro_dev.live_id = cdm.pro_dev_data_id
where cdm.id is not distinct from $1
"#,
        id.0,
    )
    .fetch_optional(pool)
    .await?;

    let map_response = |body, kind| ModuleBody::transform_response_kind(body, kind);

    match module {
        Some(it) => Ok(Some(Module {
            id: it.id,
            created_at: it.created_at,
            updated_at: it.updated_at,
            body: map_response(it.body, it.kind).context(anyhow::anyhow!(
                "failed to transform module of kind {:?}",
                it.kind
            ))?,
            is_complete: it.is_complete,
            is_updated: it.created_at < it.updated_at,
        })),
        None => Ok(None),
    }
}

/// FIXME dedup this from live JIG
pub async fn get_draft(pool: &PgPool, id: ModuleId) -> anyhow::Result<Option<Module>> {
    let module = sqlx::query!(
        //language=SQL
        r#"
select cdm.id          as "id!: ModuleId",
       contents    as "body!",
       created_at  as "created_at!",
       updated_at  as "updated_at!",
       kind        as "kind!: ModuleKind",
       is_complete as "is_complete!"
from pro_dev_data_module "cdm"
inner join pro_dev on pro_dev.draft_id = cdm.pro_dev_data_id
where cdm.id is not distinct from $1
"#,
        id.0,
    )
    .fetch_optional(pool)
    .await?;

    let map_response = |body, kind| ModuleBody::transform_response_kind(body, kind);

    match module {
        Some(it) => Ok(Some(Module {
            id: it.id,
            created_at: it.created_at,
            updated_at: it.updated_at,
            body: map_response(it.body, it.kind).context(anyhow::anyhow!(
                "failed to transform module of kind {:?}",
                it.kind
            ))?,
            is_complete: it.is_complete,
            is_updated: it.created_at < it.updated_at,
        })),
        None => Ok(None),
    }
}

pub async fn update(
    pool: &PgPool,
    parent_id: ProDevId,
    module_id: ModuleId,
    body: Option<&ModuleBody>,
    new_index: Option<u16>,
    is_complete: Option<bool>,
) -> anyhow::Result<bool> {
    let (kind, body) = match body.map(ModuleBody::map_module_contents).transpose()? {
        Some((kind, body)) => (Some(kind), Some(body)),
        None => (None, None),
    };

    let mut txn = pool.begin().await?;

    let draft_id = sqlx::query!(
        //language=SQL
        r#"
select draft_id from pro_dev where pro_dev.id = $1
"#,
        parent_id.0,
    )
    .fetch_optional(&mut txn)
    .await?
    .map(|it| it.draft_id);

    let index = sqlx::query!(
        //language=SQL
        r#"
select index from pro_dev_data_module
where pro_dev_data_id = $1 and pro_dev_data_module.id is not distinct from $2
"#,
        draft_id,
        module_id.0
    )
    .fetch_optional(&mut txn)
    .await?;

    let index = match index {
        Some(it) => it.index,
        None => return Ok(false),
    };

    sqlx::query!(
        //language=SQL
        r#"
update pro_dev_data_module
set contents    = coalesce($3, contents),
    kind        = coalesce($4, kind),
    is_complete = coalesce($5, is_complete)
where pro_dev_data_id = $1
  and index = $2
"#,
        draft_id,
        index,
        body.as_ref(),
        kind.map(|it| it as i16),
        is_complete,
    )
    .execute(&mut txn)
    .await?;

    if let Some(new_index) = new_index {
        let new_index = new_index as i16;

        let max_index = sqlx::query!(
            //language=SQL
            r#"select count(*) - 1 as "max_index!" from pro_dev_data_module where pro_dev_data_id = $1"#,
            draft_id
        )
        .fetch_one(&mut txn)
        .await?
        .max_index;

        let new_index = cmp::min(new_index, max_index as i16);

        if new_index < index {
            sqlx::query!(
                //language=SQL
                r#"
update pro_dev_data_module
set
    index = case when index = $2 then $3 else index + 1 end,
    updated_at = now()
where pro_dev_data_id = $1 and index between $3 and $2
"#,
                draft_id,
                index,
                new_index
            )
            .execute(&mut txn)
            .await?;
        } else if new_index > index {
            sqlx::query!(
                //language=SQL
                r#"
update pro_dev_data_module
set
    index = case when index = $2 then $3 else index - 1 end,
    updated_at = now()
where pro_dev_data_id = $1 and index between $2 and $3
"#,
                parent_id.0,
                index,
                new_index
            )
            .execute(&mut txn)
            .await?;
        }
    }

    txn.commit().await?;

    Ok(true)
}

pub async fn delete(pool: &PgPool, parent: ProDevId, id: ModuleId) -> anyhow::Result<()> {
    let mut txn = pool.begin().await?;

    let draft_id = sqlx::query!(
        //language=SQL
        r#"
select draft_id from pro_dev where pro_dev.id = $1
"#,
        parent.0,
    )
    .fetch_optional(&mut txn)
    .await?
    .map(|it| it.draft_id);

    let idx = sqlx::query!(
        //language=SQL
        r#"
delete
from pro_dev_data_module
where pro_dev_data_id = $1 and pro_dev_data_module.id is not distinct from $2
returning index
"#,
        draft_id,
        id.0,
    )
    .fetch_optional(&mut txn)
    .await?
    .map(|it| it.index);

    if let Some(idx) = idx {
        sqlx::query!(
            //language=SQL
            r#"
update pro_dev_data_module
set index = index - 1
where pro_dev_data_id = $1
  and index > $2
"#,
            draft_id,
            idx,
        )
        .execute(&mut txn)
        .await?;
    }

    txn.commit().await?;

    Ok(())
}
