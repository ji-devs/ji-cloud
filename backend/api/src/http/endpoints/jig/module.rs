use actix_web::{
    web::{self, Data, Json, Path, ServiceConfig},
    HttpResponse,
};
use shared::{
    api::{endpoints::jig::module, ApiEndpoint},
    domain::{
        jig::{
            module::{ModuleId, ModuleResponse, StableOrUniqueId},
            JigId,
        },
        CreateResponse,
    },
};
use sqlx::PgPool;

// use serde_qs::actix::QsQuery;

use crate::{db, error, extractor::TokenUser};

/// Create a new module on a draft JIG.
async fn create(
    db: Data<PgPool>,
    auth: TokenUser,
    parent: Path<JigId>,
    req: Json<<module::Create as ApiEndpoint>::Req>,
) -> Result<HttpResponse, error::Auth> {
    let parent_id = parent.into_inner();

    db::jig::authz(&*db, auth.0.user_id, Some(parent_id)).await?;

    let req = req.into_inner();

    let is_complete = req.body.is_complete();
    let (id, _index) = db::jig::module::create(&*db, parent_id, req.body, is_complete).await?;

    Ok(HttpResponse::Created().json(CreateResponse { id }))
}

/// Get a module from a live JIG.
async fn get_live(
    db: Data<PgPool>,
    path: web::Path<(JigId, String)>,
) -> Result<Json<<module::GetLive as ApiEndpoint>::Res>, error::NotFound> {
    let path = path.into_inner();

    let module_id = StableOrUniqueId::Unique(ModuleId(uuid::Uuid::parse_str(&path.1)?));

    let module = db::jig::module::get_live(&db, path.0, module_id)
        .await?
        .ok_or(error::NotFound::ResourceNotFound)?;

    Ok(Json(ModuleResponse { module }))
}

/// Get a module from a draft JIG.
///
/// FIXME dedup this from live JIG
async fn get_draft(
    db: Data<PgPool>,
    path: web::Path<(JigId, String)>,
) -> Result<Json<<module::GetDraft as ApiEndpoint>::Res>, error::NotFound> {
    let path = path.into_inner();

    let module_id = StableOrUniqueId::Unique(ModuleId(uuid::Uuid::parse_str(&path.1)?));

    let module = db::jig::module::get_draft(&db, path.0, module_id)
        .await?
        .ok_or(error::NotFound::ResourceNotFound)?;

    Ok(Json(ModuleResponse { module }))
}

/// Update a module in a draft JIG.
async fn update(
    db: Data<PgPool>,
    auth: TokenUser,
    path: web::Path<JigId>,
    req: Json<<module::Update as ApiEndpoint>::Req>,
) -> Result<HttpResponse, error::NotFound> {
    let parent_id = path.into_inner();

    db::jig::authz(&*db, auth.0.user_id, Some(parent_id)).await?;

    let req = req.into_inner();

    let exists = db::jig::module::update(
        &*db,
        parent_id,
        req.id,
        req.body.as_ref(),
        req.index,
        req.is_complete,
    )
    .await?;

    if !exists {
        return Err(error::NotFound::ResourceNotFound);
    }

    Ok(HttpResponse::NoContent().finish())
}

/// Delete a module from a draft JIG.
async fn delete(
    db: Data<PgPool>,
    auth: TokenUser,
    path: web::Path<JigId>,
    req: Json<<module::Delete as ApiEndpoint>::Req>,
) -> Result<HttpResponse, error::Delete> {
    let parent_id = path.into_inner();

    db::jig::authz(&*db, auth.0.user_id, Some(parent_id)).await?;

    db::jig::module::delete(&*db, parent_id, req.id).await?;

    Ok(HttpResponse::NoContent().finish())
}

pub fn configure(cfg: &mut ServiceConfig) {
    cfg.route(
        module::Create::PATH,
        module::Create::METHOD.route().to(create),
    )
    .route(
        module::GetLive::PATH,
        module::GetLive::METHOD.route().to(get_live),
    )
    .route(
        module::GetDraft::PATH,
        module::GetDraft::METHOD.route().to(get_draft),
    )
    .route(
        module::Update::PATH,
        module::Update::METHOD.route().to(update),
    )
    .route(
        module::Delete::PATH,
        module::Delete::METHOD.route().to(delete),
    );
}
