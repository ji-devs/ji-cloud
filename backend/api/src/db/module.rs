use anyhow::Context;
use shared::domain::jig::{
    module::{Module, ModuleBody, ModuleId, ModuleIdOrIndex, ModuleKind},
    JigId,
};
use sqlx::PgPool;
use std::cmp;

pub async fn create(
    pool: &PgPool,
    parent: JigId,
    body: ModuleBody,
) -> anyhow::Result<(ModuleId, u16)> {
    let (kind, body) = map_module_contents(&body)?;

    sqlx::query!(
        r#"
insert into jig_module (jig_id, kind, contents, index)
values ($1, $2, $3, (select count(*) from jig_module where jig_id = $1))
returning id, "index"
"#,
        parent.0,
        kind as i16,
        body,
    )
    .fetch_one(pool)
    .await
    .map(|it| (ModuleId(it.id), it.index as u16))
    .map_err(Into::into)
}

pub async fn update(
    pool: &PgPool,
    parent_id: JigId,
    lookup: ModuleIdOrIndex,
    body: Option<&ModuleBody>,
    new_index: Option<u16>,
    is_complete: Option<bool>,
) -> anyhow::Result<bool> {
    let (id, index) = (lookup.id(), lookup.index());

    let (kind, body) = match body.map(map_module_contents).transpose()? {
        Some((kind, body)) => (Some(kind), Some(body)),
        None => (None, None),
    };

    let mut txn = pool.begin().await?;

    let index = sqlx::query!(
        r#"select index from jig_module where jig_id = $1 and (id is not distinct from $2 or index is not distinct from $3)"#,
        parent_id.0,
        id.map(|it| it.0),
        index.map(|it| it as i16)
    )
    .fetch_optional(&mut txn)
    .await?;

    let index = match index {
        Some(it) => it.index,
        None => return Ok(false),
    };

    sqlx::query!(
        r#"
update jig_module
set contents = coalesce($3, contents),
    kind = coalesce($4, kind),
    is_complete = coalesce($5, is_complete)
where jig_id = $1 and index = $2
"#,
        parent_id.0,
        index,
        body.as_ref(),
        kind.map(|it| it as i16),
        is_complete,
    )
    .execute(&mut txn)
    .await?;

    if let Some(new_index) = new_index {
        let new_index = new_index as i16;

        // todo: don't use an extra query for this
        let max_index = sqlx::query!(
            r#"select count(*) - 1 as "max_index!" from jig_module where jig_id = $1"#,
            parent_id.0
        )
        .fetch_one(&mut txn)
        .await?
        .max_index;

        let new_index = cmp::min(new_index, max_index as i16);

        if new_index < index {
            sqlx::query!(
                r#"
update jig_module
set
    index = case when index = $2 then $3 else index + 1 end,
    updated_at = now()
where jig_id = $1 and index between $3 and $2
"#,
                parent_id.0,
                index,
                new_index
            )
            .execute(&mut txn)
            .await?;
        } else if new_index > index {
            sqlx::query!(
                r#"
update jig_module
set
    index = case when index = $2 then $3 else index - 1 end,
    updated_at = now()
where jig_id = $1 and index between $2 and $3
"#,
                parent_id.0,
                index,
                new_index
            )
            .execute(&mut txn)
            .await?;
        }
    }

    txn.commit().await?;

    Ok(true)
}

fn map_module_contents(body: &ModuleBody) -> anyhow::Result<(ModuleKind, serde_json::Value)> {
    let kind = body.kind();

    let body = match body {
        ModuleBody::CardQuiz(body) => serde_json::to_value(body)?,
        ModuleBody::Cover(body) => serde_json::to_value(body)?,
        ModuleBody::DragDrop(body) => serde_json::to_value(body)?,
        ModuleBody::Flashcards(body) => serde_json::to_value(body)?,
        ModuleBody::Matching(body) => serde_json::to_value(body)?,
        ModuleBody::MemoryGame(body) => serde_json::to_value(body)?,
        ModuleBody::Poster(body) => serde_json::to_value(body)?,
        ModuleBody::TappingBoard(body) => serde_json::to_value(body)?,
        ModuleBody::Video(body) => serde_json::to_value(body)?,

        _ => anyhow::bail!("Unimplemented body kind: {}", kind.as_str()),
    };

    Ok((kind, body))
}

fn transform_response_kind(
    contents: serde_json::Value,
    kind: ModuleKind,
) -> anyhow::Result<ModuleBody> {
    match kind {
        ModuleKind::CardQuiz => Ok(ModuleBody::CardQuiz(serde_json::from_value(contents)?)),
        ModuleKind::Cover => Ok(ModuleBody::Cover(serde_json::from_value(contents)?)),
        ModuleKind::DragDrop => Ok(ModuleBody::DragDrop(serde_json::from_value(contents)?)),
        ModuleKind::Flashcards => Ok(ModuleBody::Flashcards(serde_json::from_value(contents)?)),
        ModuleKind::Matching => Ok(ModuleBody::Matching(serde_json::from_value(contents)?)),
        ModuleKind::Memory => Ok(ModuleBody::MemoryGame(serde_json::from_value(contents)?)),
        ModuleKind::Poster => Ok(ModuleBody::Poster(serde_json::from_value(contents)?)),
        ModuleKind::TappingBoard => Ok(ModuleBody::TappingBoard(serde_json::from_value(contents)?)),
        ModuleKind::Video => Ok(ModuleBody::Video(serde_json::from_value(contents)?)),

        _ => anyhow::bail!("Unimplemented response kind"),
    }
}

pub async fn get(
    pool: &PgPool,
    parent: JigId,
    lookup: ModuleIdOrIndex,
) -> anyhow::Result<Option<Module>> {
    let (id, index) = (lookup.id(), lookup.index());

    let module = sqlx::query!(
        r#"
select 
    id as "id: ModuleId",
    contents as "body",
    kind as "kind: ModuleKind",
    is_complete as "is_complete"
from jig_module
where jig_id = $1 and (id is not distinct from $2 or index is not distinct from $3)
"#,
        parent.0,
        id.map(|it| it.0),
        index.map(|it| it as i16)
    )
    .fetch_optional(pool)
    .await?;

    let map_response = |body, kind| transform_response_kind(body, kind);

    match module {
        Some(it) => Ok(Some(Module {
            id: it.id,
            body: map_response(it.body, it.kind).context(anyhow::anyhow!(
                "failed to transform module of kind {:?}",
                it.kind
            ))?,
            is_complete: it.is_complete,
        })),
        None => Ok(None),
    }
}

pub async fn delete(pool: &PgPool, parent: JigId, lookup: ModuleIdOrIndex) -> anyhow::Result<()> {
    let (id, index) = (lookup.id(), lookup.index());
    let mut txn = pool.begin().await?;

    let idx = sqlx::query!(
        "delete from jig_module where jig_id = $1 and (id is not distinct from $2 or index is not distinct from $3) returning index",
        parent.0,
        id.map(|it| it.0),
        index.map(|it| it as i16),
    )
    .fetch_optional(&mut txn)
    .await?
    .map(|it| it.index);

    if let Some(idx) = idx {
        sqlx::query!(
            "update jig_module set index = index - 1 where jig_id = $1 and index > $2",
            parent.0,
            idx
        )
        .execute(&mut txn)
        .await?;
    }

    txn.commit().await?;

    Ok(())
}
