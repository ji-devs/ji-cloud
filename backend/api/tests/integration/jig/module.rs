use http::StatusCode;

use shared::domain::jig::module::{body::memory, ModuleBody, ModuleUpdateRequest};

use crate::{
    fixture::Fixture,
    helpers::{initialize_server, LoginExt},
};

#[actix_rt::test]
async fn update_empty() -> anyhow::Result<()> {
    let app = initialize_server(&[Fixture::User, Fixture::Jig], &[]).await;

    let port = app.port();

    let client = reqwest::Client::new();

    let resp = client
        .patch(&format!(
            "http://0.0.0.0:{}/v1/jig/0cc084bc-7c83-11eb-9f77-e3218dffb008/module/0cbfdd82-7c83-11eb-9f77-d7d86264c3bc",
            port
        ))
        .login()
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::NO_CONTENT);

    let resp = client
        .get(&format!(
            "http://0.0.0.0:{}/v1/jig/0cc084bc-7c83-11eb-9f77-e3218dffb008/module/0cbfdd82-7c83-11eb-9f77-d7d86264c3bc",
            port
        ))
        .login()
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::OK);

    let body: serde_json::Value = resp.json().await?;

    app.stop(false).await;

    insta::assert_json_snapshot!(body, {".**.updated_at" => "[timestamp]"});

    Ok(())
}

#[actix_rt::test]
async fn update_contents() -> anyhow::Result<()> {
    let app = initialize_server(&[Fixture::User, Fixture::Jig], &[]).await;

    let port = app.port();

    let client = reqwest::Client::new();

    let resp = client
        .patch(&format!(
            "http://0.0.0.0:{}/v1/jig/0cc084bc-7c83-11eb-9f77-e3218dffb008/module/0cc03a02-7c83-11eb-9f77-f77f9ad65e9a",
            port
        ))
        .login()
        .json(&ModuleUpdateRequest {
            body: Some(
                ModuleBody::MemoryGame(memory::ModuleData {
                    content: Some(memory::Content {
                        ..memory::Content::default()
                    })
                })
            ),
            is_complete: Some(true),

            ..ModuleUpdateRequest::default()
        })
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::NO_CONTENT);

    let resp = client
        .get(&format!(
            "http://0.0.0.0:{}/v1/jig/0cc084bc-7c83-11eb-9f77-e3218dffb008/module/0cc03a02-7c83-11eb-9f77-f77f9ad65e9a",
            port
        ))
        .login()
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::OK);

    let body: serde_json::Value = resp.json().await?;

    app.stop(false).await;

    insta::assert_json_snapshot!(body, {".**.updated_at" => "[timestamp]"});

    Ok(())
}
