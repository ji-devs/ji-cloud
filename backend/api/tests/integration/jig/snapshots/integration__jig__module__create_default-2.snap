---
source: tests/integration/jig/module.rs
expression: body

---
{
  "module": {
    "id": "[id]",
    "body": {
      "cover": {
        "content": null
      }
    },
    "is_complete": false,
    "is_updated": false,
    "created_at": "[created_at]",
    "updated_at": "[updated_at]"
  }
}
